File: src/controllers/auth/signup.controllers.js
import * as signupService from "../../services/auth/signup.service.js"

/**
 * @param {import("express").Request} req
 * @param {import("express").Response} res
 */
export const requestNewAccount = async (req, res) => {
  const { email } = req.body

  try {
    const resp = await signupService.requestNewAccount(email)

    if (resp.error) return res.status(400).send(resp.error)

    req.session.signup = {
        email,
        verified: false,
        verificationCode: resp.verificationCode,
        verificationCodeExpires: resp.verificationCodeExpires,
    }

    req.sessionOptions.maxAge = 60 * 60 * 1000
    req.sessionOptions.path = "/api/auth/signup/verify_email"

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const verifyEmail = async (req, res) => {
  const { code: inputCode } = req.body

  try {
    if (!req.session?.signup) return res.sendStatus(401)

    const signupSessionData = req.session.signup

    const resp = signupService.verifyEmail({ inputCode, ...signupSessionData })

    if (resp.error) return res.status(400).send(resp.error)

    req.session.signup = {
        email: signupSessionData.email,
        verified: true,
    }

    req.sessionOptions.maxAge = 60 * 60 * 1000
    req.sessionOptions.path = "/api/auth/signup/register_user"

    res.status(200).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

export const registerUser = async (req, res) => {
  try {
    if (!req.session?.signup) return res.sendStatus(401)

    const { email } = req.session.signup

    const resp = await signupService.registerUser({
      email,
      ...req.body,
    })

    if (resp.error) return res.status(400).send(resp.error)

    req.session.signup = undefined
    
    req.session.user = { authJwt: resp.jwt }

    req.sessionOptions.path = "/"
    req.sessionOptions.maxAge = 10 * 24 * 60 * 60 * 1000

    

    res.status(201).send(resp.data)
  } catch (error) {
    console.error(error)
    res.sendStatus(500)
  }
}

File: src/controllers/auth/signupControllers/signupControllers.go
package signupControllers

import (
	"i9lyfe/src/helpers"
	"i9lyfe/src/services/auth/signupService"
	"time"

	"github.com/gofiber/fiber/v3"
	"github.com/vmihailenco/msgpack/v5"
)

func RequestNewAccount(c fiber.Ctx) error {
	ctx := c.Context()

	var body requestNewAccountBody

	err := c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err := body.Validate(); err != nil {
		return err
	}

	respData, sessionData, err := signupService.RequestNewAccount(ctx, body.Email)
	if err != nil {
		return err
	}

	reqSession := map[string]any{
		"signup": sessionData,
	}

	c.Cookie(helpers.Session(reqSession, "/api/auth/signup/verify_email", int(time.Hour/time.Second)))

	return c.MsgPack(respData)
}

func VerifyEmail(c fiber.Ctx) error {
	ctx := c.Context()

	sessionData := c.Locals("signup_sess_data").(msgpack.RawMessage)

	var body verifyEmailBody

	err := c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err := body.Validate(); err != nil {
		return err
	}

	respData, newSessionData, err := signupService.VerifyEmail(ctx, sessionData, body.Code)
	if err != nil {
		return err
	}

	reqSession := map[string]any{
		"signup": newSessionData,
	}

	c.Cookie(helpers.Session(reqSession, "/api/auth/signup/register_user", int(time.Hour/time.Second)))

	return c.MsgPack(respData)
}

func RegisterUser(c fiber.Ctx) error {
	ctx := c.Context()

	sessionData := c.Locals("signup_sess_data").(msgpack.RawMessage)

	var body registerUserBody

	err := c.Bind().MsgPack(&body)
	if err != nil {
		return err
	}

	if err := body.Validate(); err != nil {
		helpers.LogError(err)
		return err
	}

	respData, authJwt, err := signupService.RegisterUser(ctx, sessionData, body.Username, body.Name, body.Bio, body.Birthday, body.Password)
	if err != nil {
		return err
	}

	reqSession := map[string]any{
		"user": map[string]any{"authJwt": authJwt},
	}

	c.Cookie(helpers.Session(reqSession, "/api/app", int(10*24*time.Hour/time.Second)))

	return c.Status(fiber.StatusCreated).MsgPack(respData)
}
